Enable multithreading in hashrows_col!

nalimilan opened the pull request:
This is somewhat faster for large number of rows (by about 20% for 10M rows and 10 threads), but a bit slower when the number of rows is lower than the threshold (so that a single thread is used). Not sure whether something can be done about it, nor whether the improvement in some cases is worth the loss in others.

```julia
julia> using Revise, DataFrames, BenchmarkTools

julia> for n in (1000, 10_000_000)
           df = DataFrame(x=rand(string.(1000000:1000010), n));
           @btime groupby($df, :x);

          # Benchmark directly the changes
           h = similar(df.x, UInt);
           n = similar(df.x, Bool);
           @btime DataFrames.hashrows_col!($h, $n, $df.x, nothing, true);
       end

# main
  21.105 μs (29 allocations: 34.05 KiB)
  12.211 μs (0 allocations: 0 bytes)
  335.166 ms (32 allocations: 280.59 MiB)
  122.653 ms (0 allocations: 0 bytes)

# PR
  26.041 μs (29 allocations: 34.05 KiB)
  16.291 μs (0 allocations: 0 bytes)
  258.437 ms (107 allocations: 280.60 MiB)
  18.463 ms (74 allocations: 5.72 KiB)
```

bkamins commented:
I was testing the same when implementing hashing in `OnCols` for joins and had similar conclusions. We probably can have it, but the threshold for turning "this on" should probably be large.

nalimilan commented:
> I was testing the same when implementing hashing in `OnCols` for joins and had similar conclusions. We probably can have it, but the threshold for turning "this on" should probably be large.

The point is, the regression happens even under the threshold. It looks like we pay the cost even when multithreading isn't used.

bkamins commented:
My guess is that we would have to split it into several methods. Maybe (I have not checked, but seen it in other cases), when you have very long methods that branch compiles is not always doing the right thing. If you will not find the root cause of this I can look at the issue later this week.

nalimilan commented:
I've switched to using a macro (inspired by `@threads` but using `@spawn`) instead of a function, and the overhead is negligible now. Actually the syntax is even nicer, it's almost identical to the non-threaded version. I guess it could replace `tforeach`.

Updated benchmark (with 4 threads):
main:
```julia
  16.520 μs (29 allocations: 34.05 KiB)
  9.406 μs (0 allocations: 0 bytes)
  207.426 ms (32 allocations: 280.59 MiB)
  100.835 ms (0 allocations: 0 bytes)
```

PR:
```julia
  17.076 μs (29 allocations: 34.05 KiB)
  10.564 μs (0 allocations: 0 bytes)
  139.823 ms (93 allocations: 280.60 MiB)
  33.247 ms (61 allocations: 5.16 KiB)
```

bkamins commented on src/groupeddataframe/utils.jl:
```diff
@@ -5,11 +5,13 @@ function hashrows_col!(h::Vector{UInt},
+        @inbounds begin
+            el = v[i]
+            h[i] = hash(el, h[i])
+            if length(n) > 0
+                n[i] |= ismissing(el)
+            end
```
Changing these lines to:
```
 if eltype(v) >: Missing && length(n) > 0
        n .|= ismissing.(v)
    end
```
test and moving it oudide might make sense in general. What do you think?

bkamins replied:
Also I have confirmed that only this line causes problem, ie. `n[i] |= ismissing(el)` does not get executed properly. I will report the details.

bkamins replied:
it happens if we work with `CategoricalVector`.

bkamins replied:
Let us keep it as is for now - it is not likely a big difference anyway

bkamins commented on src/groupeddataframe/utils.jl:
```diff
@@ -5,11 +5,13 @@ function hashrows_col!(h::Vector{UInt},
-    @inbounds for i in eachindex(h)
-        el = v[i]
-        h[i] = hash(el, h[i])
-        if length(n) > 0
```
Can you remind me why did we do `length(n) > 0` check here? (is it that `length(n) > 0` if `skipmissing` is used?)

nalimilan replied:
Yes, we pass an empty vector when `skipmissing=false`.

bkamins commented on src/other/utils.jl:
```diff
@@ -129,6 +129,61 @@ function tforeach(f, x::AbstractArray; basesize::Integer)
+                    $(esc(lbody))
+                end
+            end
+            return nothing
```
```suggestion
            nothing
```

bkamins commented on src/other/utils.jl:
```diff
@@ -129,6 +129,61 @@ function tforeach(f, x::AbstractArray; basesize::Integer)
+                    end
+                end
+            end
+            return nothing
```
```suggestion
            nothing
```

nalimilan commented:
I confirm that benchmarks give similar results with the fix.

nalimilan commented:
CI passed on the previous commit except docs, I've disabled it for the latest commit to avoid wasting resources but docs build fine locally.

bkamins commented:
Looks good now. I leave up to you if you want to investigate https://github.com/JuliaData/DataFrames.jl/pull/2685/files#r611769396 still or merge it as is.

nalimilan commented:
> Looks good now. I leave up to you if you want to investigate https://github.com/JuliaData/DataFrames.jl/pull/2685/files#r611769396 still or merge it as is.

Given that the current approach allows running both steps in parallel I'd say it's OK (and I had benchmarked it when it was introduced IIRC).
